lib/hanami/configuration.rb

```diff
@@ -3,268 +3,176 @@
-      self.sessions = DEFAULT_SESSIONS
-
-      self.router     = Router.new(base_url)
-      self.middleware = Middleware.new
-
-      self.inflections = Dry::Inflector.new
+      self.autoloader = Zeitwerk::Loader.new
 
+      # Config for actions (same for views, below) may not be available if the gem isn't
+      # loaded; fall back to a null config object if it's missing
       @actions = begin
         require_path = "hanami/action/application_configuration"
         require require_path
         Hanami::Action::ApplicationConfiguration.new
       rescue LoadError => e
         raise e unless e.path == require_path
-        Object.new
+        require_relative "configuration/null_configuration"
+        NullConfiguration.new
       end
 
+      @middleware = Middleware.new
+
+      @router = Router.new(self)
+
       @views = begin
         require_path = "hanami/view/application_configuration"
         require require_path
         Hanami::View::ApplicationConfiguration.new
       rescue LoadError => e
         raise e unless e.path == require_path
-        Object.new
-      end
-    end
-    # rubocop:enable Metrics/AbcSize, Metrics/MethodLength
-
-    def finalize
-      environment_for(env).each do |blk|
-        instance_eval(&blk)
+        require_relative "configuration/null_configuration"
+        NullConfiguration.new
       end
 
-      # Finalize nested configuration
-      #
-      # TODO: would be good to just create empty configurations for actions/views
-      #       instead of plain objects
-      actions.finalize! if actions.respond_to?(:finalize!)
-      views.finalize! if views.respond_to?(:finalize!)
-
-      self
-    end
-
-    def environment(name, &blk)
-      environment_for(name).push(blk)
-    end
-
-    def autoloader=(loader)
-      settings[:autoloader] = loader || nil
-    end
-
-    def autoloader
-      settings.fetch(:autoloader)
-    end
-
-    def env=(value)
-      settings[:env] = value
-    end
-
-    def env
-      settings.fetch(:env)
-    end
-
-    def root=(root)
-      settings[:root] = Pathname(root)
-    end
-
-    def root
-      settings.fetch(:root)
-    end
-
-    def slices_dir=(dir)
-      settings[:slices_dir] = dir
-    end
-
-    def slices_dir
-      settings.fetch(:slices_dir)
+      yield self if block_given?
     end
 
-    def slices_namespace=(namespace)
-      settings[:slices_namespace] = namespace
-    end
+    def environment(env_name, &block)
+      environments[env_name] << block
+      apply_env_config
 
-    def slices_namespace
-      settings.fetch(:slices_namespace) { Object }
+      self
     end
 
-    def slice(slice_name, &block)
-      settings[:slices][slice_name] = block
-    end
+    def finalize!
+      apply_env_config
 
-    def slices
-      settings[:slices]
-    end
+      # Finalize nested configurations
+      actions.finalize!
+      views.finalize!
+      logger.finalize!
+      router.finalize!
```

Comment:
It's becoming quite a list of things we're finalizing now! At some point we might need to take a more dynamic approach to doing this, but I think we can defer this until the point where we let plugins register their own configs, etc., which will be a whole other kettle of fish.

Reply:
@timriley I'm ok with this being hardcoded but eventually I think we want a feature-detection mechanism and everything should become a registered feature. Then this code will become something like `features.each(&:finalize)` or even `features.finalize!`.

Reply:
ps. I think I'll move logger out into a component btw, like I mentioned on multiple occasions 🙂

Reply:
@solnic Forgive me for asking given you've likely mentioned this in a place where I've already seen it (😅), but what do you mean by moving it into a component, exactly?

We do already have `lib/hanami/application/container/boot/logger.rb` which is the framework-provided bootable component for initialising and registering the logger in the Hanami application container. As of this PR, it currently looks like this:

```ruby
# frozen_string_literal: true

Hanami.application.register_bootable :logger do
  start do
    logger_config = Hanami.application.configuration.logger

    logger =
      if (logger_instance = logger_config.instance)
        logger_instance
      else
        logger_config.logger_class.new(**logger_config.options)
      end

    register :logger, logger
  end
end

```

Reply:
@timriley oh I missed your comment sorry! I thought we still use the logger plugin from dry-system to set up a logger, that's what I was referring to.